lib/spack/spack/solver/asp.py

```diff
@@ -613,8 +613,11 @@ def pkg_rules(self, pkg, tests):
         self.gen.newline()
 
         # variants
-        for name, variant in sorted(pkg.variants.items()):
-            self.gen.fact(fn.variant(pkg.name, name))
+        for name, entry in sorted(pkg.variants.items()):
+            variant, when = entry
+            for w in when:
+                cond_id = self.condition(w, name=pkg.name)
+                self.gen.fact(fn.variant_condition(pkg.name, name, cond_id))
```

Comment:
Other `*_condition` facts:
```suggestion
                cond_id = self.condition(w, name=pkg.name)
                self.gen.fact(fn.variant_condition(cond_id, pkg.name, name))
```
put the ID first.